File: python/shared/dtos/preset_member.py
from . import BaseDTO, BaseEntityDTO, NullableUrlStr
from .member import MemberDetailDTO
from .preset_member_position import PresetMemberPositionDetailDTO
from .tier import TierDTO


class PresetMemberDTO(BaseEntityDTO):
    preset_member_id: int
    preset_id: int
    member_id: int
    tier_id: int | None
    is_leader: bool


class PresetMemberDetailDTO(PresetMemberDTO):
    member: MemberDetailDTO
    tier: TierDTO | None
    preset_member_positions: list[PresetMemberPositionDetailDTO]
    info_url: str | None


class AddPresetMemberDTO(BaseDTO):
    member_id: int
    tier_id: int | None
    is_leader: bool
    info_url: NullableUrlStr


class UpdatePresetMemberDTO(BaseDTO):
    tier_id: int | None = None
    is_leader: bool | None = None
    info_url: NullableUrlStr = None

File: frontend/src/dtos/presetMember.ts
import { z } from "zod";
import { nullableUrlSchema } from "@utils/dto";
import type { MemberDetailDTO } from "./member";
import type { PresetMemberPositionDetailDTO } from "./presetMemberPosition";
import type { TierDTO } from "./tier";

export interface PresetMemberDTO {
  presetMemberId: number;
  presetId: number;
  memberId: number;
  tierId: number | null;
  isLeader: boolean;
  infoUrl: string | null;
}

export interface PresetMemberDetailDTO extends PresetMemberDTO {
  member: MemberDetailDTO;
  tier: TierDTO | null;
  presetMemberPositions: PresetMemberPositionDetailDTO[];
}

export const AddPresetMemberSchema = z.object({
  infoUrl: nullableUrlSchema,
});

export type AddPresetMemberDTO = z.infer<typeof AddPresetMemberSchema> & {
  memberId: number;
  tierId: number | null;
  isLeader: boolean;
};

export const UpdatePresetMemberSchema = AddPresetMemberSchema.partial();

export type UpdatePresetMemberDTO = z.infer<typeof UpdatePresetMemberSchema> & {
  tierId?: number | null;
  isLeader?: boolean;
};
